File: 0138. Copy List with Random Pointer/Copy List with Random Pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;

    Node() {}

    Node(int _val, Node* _next, Node* _random) {
        val = _val;
        next = _next;
        random = _random;
    }
};
*/
class Solution {
public:
    Node *copyRandomList(Node *head) {
        if (!head) return nullptr;
        Node *cur = head;
        
        while(cur) {
            Node *tmp = cur->next;
            cur->next = new Node(cur->val, nullptr, nullptr);
            cur->next->next = tmp;
            cur = tmp;
        }
        
        cur = head;
        while(cur) {
            if (cur->random) 
                cur->next->random = cur->random->next;
            cur = cur->next->next;
        }
        
        cur = head;
        Node *cCur = cur->next, *nHead = cur->next;
        while(cur) {
            cur->next = cur->next->next;
            cur = cur->next;
            cCur->next = cur ? cur->next : nullptr;
            cCur = cCur->next;
        }
        
        return nHead;
    }
};

File: 0138. Copy List with Random Pointer/Copy List with Random Pointer.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node next;
    public Node random;

    public Node() {}

    public Node(int _val,Node _next,Node _random) {
        val = _val;
        next = _next;
        random = _random;
    }
};
*/
class Solution {
    public Node copyRandomList(Node head) {
        if (head == null) return null;
        Node old = head;
        
        while (old != null) {
            Node tmp = old.next;
            old.next = new Node(old.val);
            old.next.next = tmp;
            old = tmp;
        }
        
        old = head;
        
        while (old != null) {
            if (old.random != null) old.next.random = old.random.next;
            old = old.next.next;
        }
        
        Node res = head.next, cur = head;
        old = head.next;
        
        while (cur != null) {
            cur.next = cur.next.next;
            cur = cur.next;
            
            old.next = cur != null ? cur.next : null;
            old = old.next;
        }
        
        return res;
    }
}